### ISSUE TITLE
MongoDB performance issues with `some`

### ISSUE THREAD

**Issue by masterbater** [prisma/prisma#21453]
Mongodb has some serious performance issues, that is blocking for real production db. This is really blocking some real production load
Prisma has a really good developer experience but for mongodb it seems the features and improvements are second thoughts than from SQL. I hope you inform us if  you decided to drop support for mongodb. Prisma team is doing great, I just hope you could give some attention to mongodb users.

### some
```
  where: {
            logs: { some: {} },
          },
```
<img width="269" alt="image" src="https://github.com/prisma/prisma/assets/26339988/6f50ef04-2d5b-488f-9e03-c61d7f30d0af">

**Comment by janpio** [comment]
Can you explain a bit more what we are looking at? I don't understand what the image is showing.

**Comment by janpio** [comment]
This is currently not prioritized on our roadmap, sorry. If this is blocking you, for now you'll have to find another workaround. We will update you all here when we tackle this.

**Comment by masterbater** [comment]
> This is currently not prioritized on our roadmap, sorry. If this is blocking you, for now you'll have to find another workaround. We will update you all here when we tackle this.

Is it very difficult for others to contribute since you assume that prisma team can only improve this?, also is there no dedicated team for mongodb only? I know you have a lot on your plate and you dont want to get behind. Prisma really change and make a big improvements with working with database. I hope those performance improvements is not really that far

**Comment by millsp** [comment]
Hey @masterbater could you please provide us with a reproduction?

**Comment by SevInf** [comment]
Hey @masterbater  ,

Did you get a chance to have a look at @millsp previous comment? It'd be helpful for us to move your issue forward.

Without further information, we'll have to close this issue. Thank you 🙏

**Comment by macrozone** [comment]
even on a simple findUnique, prisma leads to a colscan on the DB (!). 

I seriously think prisma team should either fix this immediatly or deprecate the mongodb support or flag it as experimental (hide it behind preview flag). 

it is NOT production ready.

**Comment by boredland** [comment]
We manually got rid of all `in` places and use `OR` for now. It is a bit verbose, but does the job.

**Comment by macrozone** [comment]
> We manually got rid of all `in` places and use `OR` for now. It is a bit verbose, but does the job. I primarily introduced prisma as a lever to migrate to pg eventually - little hacks like these are to be reverted then, I guess.

similar story, had an old meteor app where we created a new api using prisma. That worked very nicely. Just very disappointed now to see those serious performance problem which should actually be easy to fix on prisma. I mean how complicated can it be to implement a `findUnique` query against a mongodb?

**Comment by masterbater** [comment]
Any update to this, please at least give mongodb a dev time.  I see you updated the delete but the select, and includes and in queries causes to do COLSCAN instead of using mongodb index

**Comment by macrozone** [comment]
@jkomyno you added the bug labels, could you also add performance labels? since that is anyway a current roadmap topic

**Comment by jkomyno** [comment]
> @jkomyno you added the bug labels, could you also add performance labels? since that is anyway a current roadmap topic

Hi @macrozone, yes, there's already the `topic: performance` label applied 👍🏻

**Comment by boredland** [comment]
How can this have the `unconfirmed` label? Do you need more proof this bug exists?

**Comment by janpio** [comment]
No one from the team has confirmed the issue by running the provided code himself and seen the described problem, that simple. And please make sure you manage your tone @boredland. Bugs and problems are frustrating, but we are also just humans working on this. As soon as someone had the time to try to reproduce this, the label will be further updated to indicate that we could confirm the problem or not.

**Comment by masterbater** [comment]
> No one from the team has confirmed the issue by running the provided code himself and seen the described problem, that simple. And please make sure you manage your tone @boredland. Bugs and problems are frustrating, but we are also just humans working on this. As soon as someone had the time to try to reproduce this, the label will be further updated to indicate that we could confirm the problem or not.
> 
> 

You can generate two sets of collection( that is in relation to each other )with random data should be at least 10000 data, then using include. 

Then you will get an email from atlas

Query Targeting: Scanned Objects / Returned has gone above 1000
 
The ratio of documents scanned to returned exceeded 1000.0 on cluster, which typically suggests that un-indexed queries are being run.

**Comment by janpio** [comment]
None of what you write changes that this issue has not been confirmed according to our internal process, which is why it has and keeps having the `unconfirmed` label.

I will not start any further discussion about our prioritization here. This issue exists, it is labelled as a MongoDB and performance problem and we are aware of it. Software has bugs, sometimes severe ones. We are doing our job in fixing them in the order we decide based on the needs of our company, team - and users. I know this is frustrating for you right now, who are experiencing a concrete problem, that is currently not actively being worked on. But I can't just change that. And there is nothing really to tell you that will make you feel better - until we get to work on this.

By the way: Prisma is Open Source, so if you have a suggestion on how to fix this problem, we would of course be happy about Pull Requests to fix our implementation. Just to make that explicit.

(And as a more personal side note: Being not nice to the maintainers and people that are talking to you (instead of closing issues, or just ignore them silently), and probably advocating for you internally, is usually not the best strategy to get what you want. I have been doing this for a while, so don't get overly emotional in these discussions - but others in the team have not and are not really excited to look into these issues when the discussion is getting unpleasant. Please keep that in mind.)

**Comment by masterbater** [comment]
> None of what you write changes that this issue has not been confirmed according to our internal process, which is why it has and keeps having the `unconfirmed` label.
> 
> I will not start any further discussion about our prioritization here. This issue exists, it is labelled as a MongoDB and performance problem and we are aware of it. Software has bugs, sometimes severe ones. We are doing our job in fixing them in the order we decide based on the needs of our company, team - and users. I know this is frustrating for you right now, who are experiencing a concrete problem, that is currently not actively being worked on. But I can't just change that. And there is nothing really to tell you that will make you feel better - until we get to work on this.
> 
> By the way: Prisma is Open Source, so if you have a suggestion on how to fix this problem, we would of course be happy about Pull Requests to fix our implementation. Just to make that explicit.
> 
> (And as a more personal side note: Being not nice to the maintainers and people that are talking to you (instead of closing issues, or just ignore them silently), and probably advocating for you internally, is usually not the best strategy to get what you want. I have been doing this for a while, so don't get overly emotional in these discussions - but others in the team have not and are not really excited to look into these issues when the discussion is getting unpleasant. Please keep that in mind.)

I think you are being a little sensitive and stress, they provide constructive criticism and feedback and opinions for prisma (No one is trolling and directly attacking or insulting maintainers or your team).  Prisma team built a great software and easy and intuitive syntax. We use these it makes our development 10 times faster and safer because of the typesafety. You have great milestone for performance, prisma 5 json protocol, relatedJoins etc, but unfortunately mongodb is not a priority for Prisma team right now, we even considering migrating our mongodb to postgresql just so we can continue to use prisma, thats how we love prisma. Hopefully this severe performance issue will get your attention not so far in the future.

**Comment by boredland** [comment]
> But I can't just change that.

What you could & should do as a serious open-source maintainer with not enough time on his hands - and nothing more has been asked multiple times - is to re-flag mongodb-support of prisma as in alpha/preview for the time being, so people don't start projects with prisma to then realize that they're trapped in a library that has severe performance issues with even the simplest of queries.

**Comment by macrozone** [comment]
i also have found a `some` query which leads to another colscan

```

// prisma.schema


model Company {
  id                   String                    @id @map("_id")
  // ... 
  participations       CompaniesParticipations[]

  @@map("companies")
}

type CompaniesParticipations {
  campaignId    String
  createdAt     DateTime?                           @db.Date
  employeeCount Int
  months        String
  //....
}


// this query 
await prisma.company.count({
          where: {
            participations: {
              some: {
                campaignId: campaign.id,
              },
            },
          },
        });
```

this creates an overly complicated query:

```
"$match" : {
                        "$expr" : {
                            "$and" : [
                                {
                                    "$gt" : [
                                        {
                                            "$size" : {
                                                "$filter" : {
                                                    "input" : {
                                                        "$ifNull" : [
                                                            "$participations",
                                                            [

                                                            ]
                                                        ]
                                                    },
                                                    "as" : "elem",
                                                    "cond" : {
                                                        "$and" : [
                                                            {
                                                                "$eq" : [
                                                                    "$$elem.campaignId",
                                                                    {
                                                                        "$literal" : "EuFvdM5yy747QG6yP"
                                                                    }
                                                                ]
                                                            },
                                                            {
                                                                "$ne" : [
                                                                    "$$elem.campaignId",
                                                                    "$$REMOVE"
                                                                ]
                                                            }
                                                        ]
                                                    }
                                                }
                                            }
                                        },
                                        NumberInt(0)
                                    ]
                                },
                                {
                                    "$ne" : [
                                        "$participations",
                                        "$$REMOVE"
                                    ]
                                }
                            ]
                        }
                    }
                },
                {
                    "$group" : {
                        "_id" : null,
                        "count_all" : {
                            "$sum" : NumberInt(1)
                        }
                    }
                }
            ],
```

this bypasses any index. The correct query would be trivial in mongodb.

**Comment by janpio** [comment]
@macrozone I split out your reported problem to its own issue https://github.com/prisma/prisma/issues/24034, as we realized this is about composite types / embedded documents and unrelated to what was reported here.

**Comment by janpio** [comment]
I have redefined this issue here to only be about the performance problem caused by `some` as originally reported in the issue description by @masterbater. (We have enough other issues about `include` and `in`, see for example https://github.com/prisma/prisma/issues/12793 or https://github.com/prisma/prisma/issues/15156).

Please keep the discussion in this issue thread to the general problem with `some` going forward. Thank you.

**Comment by janpio** [comment]
@masterbater, we tried to reproduce the problem with `some` that you reported. I am quoting a colleague from an internal document:

> I used the following schema and TS script for testing:
> 
> <details>
> <summary>Schema</summary>
> 
> ```prisma
> generator client {
>   provider        = "prisma-client-js"
>   previewFeatures = ["fullTextSearch"]
> }
> 
> datasource db {
>   provider = "mongodb"
>   url      = "mongodb://localhost:27017/test"
> }
> 
> model Test {
>   id      String @id @default(auto()) @map("_id") @db.ObjectId
>   content String
> }
> 
> model User {
>   id       String @id @default(auto()) @map("_id") @db.ObjectId
>   username String
>   post     Post[]
> }
> 
> model Post {
>   id       String @id @default(auto()) @map("_id") @db.ObjectId
>   content  String
>   authorId String @db.ObjectId
>   author   User   @relation(fields: [authorId], references: [id])
> 
>   @@index([authorId])
> }
> ```
> </details>
> 
> <details>
> <summary>TS Script</summary>
> 
> ```ts
> import { PrismaClient } from '@prisma/client'
> 
> const prisma = new PrismaClient({
>     log: ['query']
> })
> 
> const RECORD_COUNT = 100_000;
> 
> async function seed() {
>     await prisma.user.deleteMany({});
>     await prisma.post.deleteMany({});
> 
>     for (let i = 0; i < RECORD_COUNT; i++) {
>         await prisma.user.create({
>             data: {
>                 username: "laplab",
>                 post: {
>                     create: {
>                         content: "content"
>                     }
>                 }
>             }
>         })
>     }
> }
> 
> async function main() {
>     await seed();
> 
>     const userCount = await prisma.user.count({});
>     const postCount = await prisma.post.count({});
>     console.log(`Database has ${userCount} users and ${postCount} posts.`);
> 
>     await prisma.user.findMany({
>         where: {
>             post: {
>                 some: {}
>             }
>         }
>     });
> }
> 
> main()
>     .then(async () => {
>         await prisma.$disconnect()
>     })
>     .catch(async (e) => {
>         console.error(e)
>         await prisma.$disconnect()
>         process.exit(1)
>     })
> ```
> </details>
> 
> Querying with `some` operator produces the following MongoDB query:
> 
> <details>
> <summary>MongoDB query</summary>
> 
> ```json
> [
>   {
>     $lookup: {
>       from: "Post",
>       let: { left_0: "$_id" },
>       pipeline: [
>         {
>           $match: {
>             $expr: {
>               $and: [
>                 {
>                   $eq: ["$authorId", "$$left_0"],
>                 },
>               ],
>             },
>           },
>         },
>       ],
>       as: "PostToUser",
>     },
>   },
>   {
>     $match: {
>       $expr: {
>         $gt: [
>           {
>             $size: {
>               $filter: {
>                 input: "$PostToUser",
>                 as: "elem",
>                 cond: {},
>               },
>             },
>           },
>           0,
>         ],
>       },
>     },
>   },
>   { $project: { _id: 1, username: 1 } },
> ]
> ```
> </details>
> 
> MongoDB Compass produces a very confusing explain visualisation, which might lead some users to believe that we do not use any indexes at all:
> 
> ![CleanShot 2024-04-30 at 14 12 20@2x](https://github.com/prisma/team-orm/assets/5737185/b5931b5c-2a1c-4ab8-b861-319f5315042f)
> 
> It even outright says that no index is available for the query. This, however, is completely untrue. The part of the query which does not use index is the scan of `User` collection. The query literally requires us to scan the whole collection without any direct filtering, there is no index to be used here. However, the rest of the query (most importantly, `$lookup`) is not present in the visualisation at all.
> 
> Here is a complete JSON output of explain command issued on this query:
> 
> <details>
> <summary>Explain output</summary>
> 
> ```json
> {
>   "explainVersion": "1",
>   "stages": [
>     {
>       "$cursor": {
>         "queryPlanner": {
>           "namespace": "test.User",
>           "indexFilterSet": false,
>           "parsedQuery": {},
>           "queryHash": "14756AA4",
>           "planCacheKey": "14756AA4",
>           "maxIndexedOrSolutionsReached": false,
>           "maxIndexedAndSolutionsReached": false,
>           "maxScansToExplodeReached": false,
>           "winningPlan": {
>             "stage": "PROJECTION_SIMPLE",
>             "transformBy": {
>               "_id": 1,
>               "username": 1
>             },
>             "inputStage": {
>               "stage": "COLLSCAN",
>               "direction": "forward"
>             }
>           },
>           "rejectedPlans": []
>         },
>         "executionStats": {
>           "executionSuccess": true,
>           "nReturned": 29982,
>           "executionTimeMillis": 724,
>           "totalKeysExamined": 0,
>           "totalDocsExamined": 29982,
>           "executionStages": {
>             "stage": "PROJECTION_SIMPLE",
>             "nReturned": 29982,
>             "executionTimeMillisEstimate": 1,
>             "works": 29983,
>             "advanced": 29982,
>             "needTime": 0,
>             "needYield": 0,
>             "saveState": 32,
>             "restoreState": 32,
>             "isEOF": 1,
>             "transformBy": {
>               "_id": 1,
>               "username": 1
>             },
>             "inputStage": {
>               "stage": "COLLSCAN",
>               "nReturned": 29982,
>               "executionTimeMillisEstimate": 1,
>               "works": 29983,
>               "advanced": 29982,
>               "needTime": 0,
>               "needYield": 0,
>               "saveState": 32,
>               "restoreState": 32,
>               "isEOF": 1,
>               "direction": "forward",
>               "docsExamined": 29982
>             }
>           },
>           "allPlansExecution": []
>         }
>       },
>       "nReturned": 29982,
>       "executionTimeMillisEstimate": 30
>     },
>     {
>       "$lookup": {
>         "from": "Post",
>         "as": "PostToUser",
>         "let": { "left_0": "$_id" },
>         "pipeline": [
>           {
>             "$match": {
>               "$expr": {
>                 "$and": [
>                   {
>                     "$eq": [
>                       "$authorId",
>                       "$$left_0"
>                     ]
>                   }
>                 ]
>               }
>             }
>           }
>         ]
>       },
>       "totalDocsExamined": 29982,
>       "totalKeysExamined": 29982,
>       "collectionScans": 0,
>       "indexesUsed": ["Post_authorId_idx"],
>       "nReturned": 29982,
>       "executionTimeMillisEstimate": 724
>     },
>     {
>       "$match": {
>         "$expr": {
>           "$gt": [
>             {
>               "$size": [
>                 {
>                   "$filter": {
>                     "input": "$PostToUser",
>                     "as": "elem",
>                     "cond": { "$const": {} }
>                   }
>                 }
>               ]
>             },
>             { "$const": 0 }
>           ]
>         }
>       },
>       "nReturned": 29982,
>       "executionTimeMillisEstimate": 724
>     },
>     {
>       "$project": {
>         "_id": true,
>         "username": true
>       },
>       "nReturned": 29982,
>       "executionTimeMillisEstimate": 724
>     }
>   ],
>   "serverInfo": {
>     "host": "MacBook-Pro.fritz.box",
>     "port": 27017,
>     "version": "7.0.8",
>     "gitVersion": "c5d33e55ba38d98e2f48765ec4e55338d67a4a64"
>   },
>   "serverParameters": {
>     "internalQueryFacetBufferSizeBytes": 104857600,
>     "internalQueryFacetMaxOutputDocSizeBytes": 104857600,
>     "internalLookupStageIntermediateDocumentMaxSizeBytes": 104857600,
>     "internalDocumentSourceGroupMaxMemoryBytes": 104857600,
>     "internalQueryMaxBlockingSortMemoryUsageBytes": 104857600,
>     "internalQueryProhibitBlockingMergeOnMongoS": 0,
>     "internalQueryMaxAddToSetBytes": 104857600,
>     "internalDocumentSourceSetWindowFieldsMaxMemoryBytes": 104857600,
>     "internalQueryFrameworkControl": "trySbeRestricted"
>   },
>   "command": {
>     "aggregate": "User",
>     "pipeline": [
>       {
>         "$lookup": {
>           "from": "Post",
>           "let": { "left_0": "$_id" },
>           "pipeline": [
>             {
>               "$match": {
>                 "$expr": {
>                   "$and": [
>                     {
>                       "$eq": [
>                         "$authorId",
>                         "$$left_0"
>                       ]
>                     }
>                   ]
>                 }
>               }
>             }
>           ],
>           "as": "PostToUser"
>         }
>       },
>       {
>         "$match": {
>           "$expr": {
>             "$gt": [
>               {
>                 "$size": {
>                   "$filter": {
>                     "input": "$PostToUser",
>                     "as": "elem",
>                     "cond": {}
>                   }
>                 }
>               },
>               0
>             ]
>           }
>         }
>       },
>       { "$project": { "_id": 1, "username": 1 } }
>     ],
>     "cursor": {},
>     "maxTimeMS": 60000,
>     "$db": "test"
>   },
>   "ok": 1,
>   "$clusterTime": {
>     "clusterTime": {
>       "$timestamp": "7363647272086667265"
>     },
>     "signature": {
>       "hash": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=",
>       "keyId": 0
>     }
>   },
>   "operationTime": {
>     "$timestamp": "7363647272086667265"
>   }
> }
> ```
> </details>
> 
> If we take a look at the part relevant to the `$lookup`, we can see:
> 
> ```json
> {
>       "$lookup": {
>         "from": "Post",
>         "as": "PostToUser",
>         "let": { "left_0": "$_id" },
>         "pipeline": [ ... ]
>       },
>       "totalDocsExamined": 29982,
>       "totalKeysExamined": 29982,
>       "collectionScans": 0,
>       "indexesUsed": ["Post_authorId_idx"],
>       "nReturned": 29982,
>       "executionTimeMillisEstimate": 724
>     },
> ```
> 
> Note the `indexesUsed` key here. `$lookup` operator was able to use the index we have defined on the `Post` collection without any issues.

To summarize in my words: It seems we can not reproduce any performance problem caused by the simple usage of `some` in a Prisma Client query similar to the snippet you had shared @masterbater. 

We did our best to come up with a full query and the needed Prima schema to reproduce. Did we make a mistake somewhere and miss important information that would allow us to reproduce the problem you described?

**Comment by masterbater** [comment]
> @masterbater, we tried to reproduce the problem with `some` that you reported. I am quoting a colleague from an internal document:
> 
> > I used the following schema and TS script for testing:
> > 
> > Schema
> > ```prisma
> > generator client {
> >   provider        = "prisma-client-js"
> >   previewFeatures = ["fullTextSearch"]
> > }
> > 
> > datasource db {
> >   provider = "mongodb"
> >   url      = "mongodb://localhost:27017/test"
> > }
> > 
> > model Test {
> >   id      String @id @default(auto()) @map("_id") @db.ObjectId
> >   content String
> > }
> > 
> > model User {
> >   id       String @id @default(auto()) @map("_id") @db.ObjectId
> >   username String
> >   post     Post[]
> > }
> > 
> > model Post {
> >   id       String @id @default(auto()) @map("_id") @db.ObjectId
> >   content  String
> >   authorId String @db.ObjectId
> >   author   User   @relation(fields: [authorId], references: [id])
> > 
> >   @@index([authorId])
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 
> > TS Script
> > ```ts
> > import { PrismaClient } from '@prisma/client'
> > 
> > const prisma = new PrismaClient({
> >     log: ['query']
> > })
> > 
> > const RECORD_COUNT = 100_000;
> > 
> > async function seed() {
> >     await prisma.user.deleteMany({});
> >     await prisma.post.deleteMany({});
> > 
> >     for (let i = 0; i < RECORD_COUNT; i++) {
> >         await prisma.user.create({
> >             data: {
> >                 username: "foo",
> >                 post: {
> >                     create: {
> >                         content: "content"
> >                     }
> >                 }
> >             }
> >         })
> >     }
> > }
> > 
> > async function main() {
> >     await seed();
> > 
> >     const userCount = await prisma.user.count({});
> >     const postCount = await prisma.post.count({});
> >     console.log(`Database has ${userCount} users and ${postCount} posts.`);
> > 
> >     await prisma.user.findMany({
> >         where: {
> >             post: {
> >                 some: {}
> >             }
> >         }
> >     });
> > }
> > 
> > main()
> >     .then(async () => {
> >         await prisma.$disconnect()
> >     })
> >     .catch(async (e) => {
> >         console.error(e)
> >         await prisma.$disconnect()
> >         process.exit(1)
> >     })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > Querying with `some` operator produces the following MongoDB query:
> > 
> > MongoDB query
> > ```json
> > [
> >   {
> >     $lookup: {
> >       from: "Post",
> >       let: { left_0: "$_id" },
> >       pipeline: [
> >         {
> >           $match: {
> >             $expr: {
> >               $and: [
> >                 {
> >                   $eq: ["$authorId", "$$left_0"],
> >                 },
> >               ],
> >             },
> >           },
> >         },
> >       ],
> >       as: "PostToUser",
> >     },
> >   },
> >   {
> >     $match: {
> >       $expr: {
> >         $gt: [
> >           {
> >             $size: {
> >               $filter: {
> >                 input: "$PostToUser",
> >                 as: "elem",
> >                 cond: {},
> >               },
> >             },
> >           },
> >           0,
> >         ],
> >       },
> >     },
> >   },
> >   { $project: { _id: 1, username: 1 } },
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > MongoDB Compass produces a confusing explain visualisation, which might lead some users to believe that we do not use any indexes at all:
> > ![CleanShot 2024-04-30 at 14 12 20@2x](https://github.com/prisma/team-orm/assets/5737185/b5931b5c-2a1c-4ab8-b861-319f5315042f)
> > It even outright says that no index is available for the query. This, however, is completely untrue. The part of the query which does not use index is the scan of `User` collection. The query literally requires us to scan the whole collection without any direct filtering, there is no index to be used here of course. However, the rest of the query (most importantly, `$lookup`) is not present in the visualisation at all.
> > Here is a complete JSON output of explain command issued on this query:
> > 
> > Explain output
> > ```json
> > {
> >   "explainVersion": "1",
> >   "stages": [
> >     {
> >       "$cursor": {
> >         "queryPlanner": {
> >           "namespace": "test.User",
> >           "indexFilterSet": false,
> >           "parsedQuery": {},
> >           "queryHash": "14756AA4",
> >           "planCacheKey": "14756AA4",
> >           "maxIndexedOrSolutionsReached": false,
> >           "maxIndexedAndSolutionsReached": false,
> >           "maxScansToExplodeReached": false,
> >           "winningPlan": {
> >             "stage": "PROJECTION_SIMPLE",
> >             "transformBy": {
> >               "_id": 1,
> >               "username": 1
> >             },
> >             "inputStage": {
> >               "stage": "COLLSCAN",
> >               "direction": "forward"
> >             }
> >           },
> >           "rejectedPlans": []
> >         },
> >         "executionStats": {
> >           "executionSuccess": true,
> >           "nReturned": 29982,
> >           "executionTimeMillis": 724,
> >           "totalKeysExamined": 0,
> >           "totalDocsExamined": 29982,
> >           "executionStages": {
> >             "stage": "PROJECTION_SIMPLE",
> >             "nReturned": 29982,
> >             "executionTimeMillisEstimate": 1,
> >             "works": 29983,
> >             "advanced": 29982,
> >             "needTime": 0,
> >             "needYield": 0,
> >             "saveState": 32,
> >             "restoreState": 32,
> >             "isEOF": 1,
> >             "transformBy": {
> >               "_id": 1,
> >               "username": 1
> >             },
> >             "inputStage": {
> >               "stage": "COLLSCAN",
> >               "nReturned": 29982,
> >               "executionTimeMillisEstimate": 1,
> >               "works": 29983,
> >               "advanced": 29982,
> >               "needTime": 0,
> >               "needYield": 0,
> >               "saveState": 32,
> >               "restoreState": 32,
> >               "isEOF": 1,
> >               "direction": "forward",
> >               "docsExamined": 29982
> >             }
> >           },
> >           "allPlansExecution": []
> >         }
> >       },
> >       "nReturned": 29982,
> >       "executionTimeMillisEstimate": 30
> >     },
> >     {
> >       "$lookup": {
> >         "from": "Post",
> >         "as": "PostToUser",
> >         "let": { "left_0": "$_id" },
> >         "pipeline": [
> >           {
> >             "$match": {
> >               "$expr": {
> >                 "$and": [
> >                   {
> >                     "$eq": [
> >                       "$authorId",
> >                       "$$left_0"
> >                     ]
> >                   }
> >                 ]
> >               }
> >             }
> >           }
> >         ]
> >       },
> >       "totalDocsExamined": 29982,
> >       "totalKeysExamined": 29982,
> >       "collectionScans": 0,
> >       "indexesUsed": ["Post_authorId_idx"],
> >       "nReturned": 29982,
> >       "executionTimeMillisEstimate": 724
> >     },
> >     {
> >       "$match": {
> >         "$expr": {
> >           "$gt": [
> >             {
> >               "$size": [
> >                 {
> >                   "$filter": {
> >                     "input": "$PostToUser",
> >                     "as": "elem",
> >                     "cond": { "$const": {} }
> >                   }
> >                 }
> >               ]
> >             },
> >             { "$const": 0 }
> >           ]
> >         }
> >       },
> >       "nReturned": 29982,
> >       "executionTimeMillisEstimate": 724
> >     },
> >     {
> >       "$project": {
> >         "_id": true,
> >         "username": true
> >       },
> >       "nReturned": 29982,
> >       "executionTimeMillisEstimate": 724
> >     }
> >   ],
> >   "serverInfo": {
> >     "host": "MacBook-Pro.fritz.box",
> >     "port": 27017,
> >     "version": "7.0.8",
> >     "gitVersion": "c5d33e55ba38d98e2f48765ec4e55338d67a4a64"
> >   },
> >   "serverParameters": {
> >     "internalQueryFacetBufferSizeBytes": 104857600,
> >     "internalQueryFacetMaxOutputDocSizeBytes": 104857600,
> >     "internalLookupStageIntermediateDocumentMaxSizeBytes": 104857600,
> >     "internalDocumentSourceGroupMaxMemoryBytes": 104857600,
> >     "internalQueryMaxBlockingSortMemoryUsageBytes": 104857600,
> >     "internalQueryProhibitBlockingMergeOnMongoS": 0,
> >     "internalQueryMaxAddToSetBytes": 104857600,
> >     "internalDocumentSourceSetWindowFieldsMaxMemoryBytes": 104857600,
> >     "internalQueryFrameworkControl": "trySbeRestricted"
> >   },
> >   "command": {
> >     "aggregate": "User",
> >     "pipeline": [
> >       {
> >         "$lookup": {
> >           "from": "Post",
> >           "let": { "left_0": "$_id" },
> >           "pipeline": [
> >             {
> >               "$match": {
> >                 "$expr": {
> >                   "$and": [
> >                     {
> >                       "$eq": [
> >                         "$authorId",
> >                         "$$left_0"
> >                       ]
> >                     }
> >                   ]
> >                 }
> >               }
> >             }
> >           ],
> >           "as": "PostToUser"
> >         }
> >       },
> >       {
> >         "$match": {
> >           "$expr": {
> >             "$gt": [
> >               {
> >                 "$size": {
> >                   "$filter": {
> >                     "input": "$PostToUser",
> >                     "as": "elem",
> >                     "cond": {}
> >                   }
> >                 }
> >               },
> >               0
> >             ]
> >           }
> >         }
> >       },
> >       { "$project": { "_id": 1, "username": 1 } }
> >     ],
> >     "cursor": {},
> >     "maxTimeMS": 60000,
> >     "$db": "test"
> >   },
> >   "ok": 1,
> >   "$clusterTime": {
> >     "clusterTime": {
> >       "$timestamp": "7363647272086667265"
> >     },
> >     "signature": {
> >       "hash": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=",
> >       "keyId": 0
> >     }
> >   },
> >   "operationTime": {
> >     "$timestamp": "7363647272086667265"
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > If we take a look at the part relevant to the `$lookup`, we can see:
> > ```json
> > {
> >       "$lookup": {
> >         "from": "Post",
> >         "as": "PostToUser",
> >         "let": { "left_0": "$_id" },
> >         "pipeline": [ ... ]
> >       },
> >       "totalDocsExamined": 29982,
> >       "totalKeysExamined": 29982,
> >       "collectionScans": 0,
> >       "indexesUsed": ["Post_authorId_idx"],
> >       "nReturned": 29982,
> >       "executionTimeMillisEstimate": 724
> >     },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Note the `indexesUsed` key here. `$lookup` operator was able to use the index we have defined on the `Post` collection without any issues.
> 
> To summarize in my words: It seems we can not reproduce any performance problem caused by the simple usage of `some` in a Prisma Client query similar to the snippet you had shared @masterbater.
> 
> We did our best to come up with a full query and the needed Prima schema to reproduce. Did we make a mistake somewhere and miss important information that would allow us to reproduce the problem you described?

Right it did use keys examined, originally i posted both include and in since most of my query has include or select. But most of the performance issued I encountered was mitigated buy creating a view or materialized view mongodb, and use that in prisma. 

<img width="531" alt="image" src="https://github.com/prisma/prisma/assets/26339988/15ce5ace-d982-41bf-81bd-3f50a52b8c9f">

**Comment by masterbater** [comment]
But its great it did get this a needed attention, hopefully all mongodb performance issue will be fix in next version 5.14

**Comment by janpio** [comment]
So do you confirm that there is no problem with `some` and this issue can be closed @masterbater?

**Comment by janpio** [comment]
ping @masterbater for my last question

**Comment by janpio** [comment]
Note that our release 5.14.0 significantly improved the performance of certain queries when using MongoDB, specifically ones with `include` and `in` should now not cause a collection scan any more and use an index instead to find the correct documents. 

If you still encounter any performance problems with Prisma and MongoDB, please open a new GitHub bug issue. Thanks!
